### ISSUE TITLE
[2.0] clarify processing input/output

### ISSUE THREAD

**Issue by tmchartrand** [AllenNeuralDynamics/aind-data-schema#1167]
**Is your feature request related to a problem? Please describe.**
In many data processing tasks there are multiple distinct inputs, e.g. generating the smartspim template or training a model. We added minimal support for this in #1166 in the form of a list of input locations, but discussed a plan to add additional context explaining the selection of those inputs.

**Describe the solution you'd like**
```python
class CompositeData(AindModel):
    """Description of a group of data assets used together"""
    data_assets: List[str] = Field(..., title="Data assets")
    name: Optional[str]
    shared_metadata: Optional[AindGenericType] = Field(
        default=None,
        title="Shared metadata", 
        description="Common attributes that provide context for this grouping of assets"
    )
    curation_purpose: Optional[str] = Field(
        default=None,
        title="Curation purpose",
        description="Reason for grouping assets together for processing"
    )
```

**Describe alternatives you've considered**
The most obvious alternative is to create a separate DataProcess stage for "curation" or the like, with no inputs and the list of curated assets as an output parameter. Other contextual fields could be entered as notes or parameters of the process.

**Additional context**
One advantage of a separate object for this is that it would be easier to write a script to run a new process on the same set of data assets as a previous processing result (could be a common use case for model evaluation etc).
Unclear if this would be reused/allowed anywhere other than the input field of DataProcess

This is also related to the discussion https://github.com/AllenNeuralDynamics/aind-data-schema/discussions/1148

**Comment by saskiad** [comment]
I wonder if we can be more specific about the shared metadata. Not sure what that looks like, I just want to avoid this being a huge dump.

**Comment by tmchartrand** [comment]
Yes I guess we'd discussed limiting that to something like "Key shared attributes used to select this group of assets" - maybe the title could actually be "Defining metadata" or "Defining attributes"?

**Comment by tmchartrand** [comment]
@dbirman added the 2.0 label to this to keep it on your radar as something that Saskia and I thought should be coordinated with that update.

**Comment by tmchartrand** [comment]
this should probably also include a bit of upgrader work checking for existing metadata that uses a comma-separated list in the current fields like input_location etc (probably not many but at least a few)

**Comment by dbirman** [comment]
Processing input_location will become Union[DataAsset, CombinedDataAsset]

@tmchartrand @dyf what's the name going to be on CO for the combine assets?

**Comment by tmchartrand** [comment]
(closely linked to #1257 )

**Comment by dbirman** [comment]
`output_location` unclear what we actually want here

**Comment by saskiad** [comment]
> since the assets are currently de-emphasized in the schema,

I'm confused by this statement. The metadata schema describes individual assets, so I'm unclear how they are de-emphasized. I presume you mean something a little bit different?

**Comment by dbirman** [comment]
> > since the assets are currently de-emphasized in the schema,
> 
> I'm confused by this statement. The metadata schema describes individual assets, so I'm unclear how they are de-emphasized. I presume you mean something a little bit different?

I think Tom meant that we aren't explicitly representing the AIND-specific concept of data assets, but like we discussed I think we should start doing it while wrapping it in a Union so non-aind folks can still use paths to other kinds of things

**Comment by tmchartrand** [comment]
@dbirman just thought of one more consideration on the broader input_location changes: should probably be Union[DataAsset, CompositeDataAsset, ModelAsset], recognizing that model weights may be stored/accessed differently than other data (as in the CO mlflow feature)

**Comment by dbirman** [comment]
Hmm, that doesn't really make sense to me but I think it's because I haven't looked closely at the model metadata / don't really understand what's going on there. Maybe something we can look at together quickly later this week

**Comment by dbirman** [comment]
```

class Provenanace(DataModel):

    code: Code = Field(..., title="Code used for processing")
    input_location: Union[str, List[str]] = Field(..., description="Path(s) to data inputs", title="Input location")
    parameters: GenericModelType = Field(default=GenericModel(), title="Parameters")
```

output_location becomes a Path

**Comment by tmchartrand** [comment]
Tentative suggestion for updates to the Code object based on discussions of provenance with @dyf : 

- rename Code.software to Code.primary_dependency or remove (if removed but needed for provenance for a particular use this could be passed under Code.parameters)
- for use when there isn't actually any code, but just say a binary software package, either:
	- still use Code, adding required `Code.name: str = Field(...)`
	- or make the DataProcess.code field have type Code | Software (and possibly rename?)

Also, documenting our resolution of the ModelAsset point above: decided against any separate representation of Model vs Data assets since they are effectively the same from the code ocean API perspective and most other purposes.

**Comment by tmchartrand** [comment]
One additional complication for the input field is that there may be no input asset for stages that represent the middle of a pipeline (or at least the input asset wouldn't be informative). After some brief discussion with Saskia, we were leaning towards allowing the input field to also point to an intermediate result referencing another processing stage (or just reference the other processing step directly by name).
I can help by taking a stab at a PR for all related parts of this issue at some point next week if helpful.

**Comment by dbirman** [comment]
I opened a branch 1167-20-clarify-processing-inputoutput and took care of the parts that seemed straightforward on this. I'll leave it to you all to look at the rest.

Heads up that I'm still aiming to close out 2.0 tickets this week.

**Comment by dbirman** [comment]
Some notes:

- curation_purpose -> description
- Remove DataProcess.parameters (in Code)
- Maybe rename experimenters for DataProcess
- David doesn't like Software but he's wrong

**Comment by dbirman** [comment]
@tmchartrand this was part of the processing merge right?
